lib/ansible/modules/cloud/amazon/GUIDELINES.md

```diff
@@ -142,7 +142,8 @@ else:
 
 An example of connecting to ec2 is shown below.  Note that there is no `NoAuthHandlerFound`
 exception handling like in boto.  Instead, an `AuthFailure` exception will be thrown when you use
-'connection'. See exception handling.
+'connection'. To ensure that authorization, parameter validation and permissions errors are all
+caught, you should catch `ClientError` and `BotoCoreError` exceptions with every boto3 connection call.
```

Comment:
Do we know when BotoCoreError occurs rather than ClientError? I think I've only experience ClientError when providing invalid credentials.

Reply:
BotoCoreError covers ParamValidationError and NoCredentialsError

```
In [13]: inspect.getclasstree(inspect.getmro(botocore.exceptions.NoCredentialsError))
Out[13]: 
[(object, ()),
 [(BaseException, (object,)),
  [(Exception, (BaseException,)),
   [(botocore.exceptions.BotoCoreError, (Exception,)),
    [(botocore.exceptions.NoCredentialsError,
      (botocore.exceptions.BotoCoreError,))]]]]]
```

NoCredentialsError is what happens if you call any API method (e.g. client.describe_instances) without the credentials being properly passed (i.e. passing no values, rather than passing incorrect values)

With no default profile, the latter suffices to generate a NoCredentialsError:

```
import botocore
client = boto3.session.Session().client('ec2', region_name='ap-southeast-2')
client.describe_instances()
```